nexus/src/app/sagas/instance_create.rs

```diff
@@ -1021,6 +1036,90 @@ async fn sic_delete_instance_record(
     Ok(())
 }
 
+// This is done intentionally late in instance creation:
+// * if the boot disk is provided by name and that disk is created along with
+//   the disk, there would not have been an ID to use any earlier
+// * if the boot disk is pre-existing, we still must wait for `disk-attach`
+//   subsagas to complete; attempting to set the boot disk earlier would error
+//   out because the desired boot disk is not attached.
+/// Set the instance's boot disk, if one was specified.
+async fn sic_set_boot_disk(
+    sagactx: NexusActionContext,
+) -> Result<(), ActionError> {
+    let osagactx = sagactx.user_data();
+    let params = sagactx.saga_params::<Params>()?;
+    let datastore = osagactx.datastore();
+    let opctx = crate::context::op_context_for_saga_action(
+        &sagactx,
+        &params.serialized_authn,
+    );
+
+    // TODO: instead of taking this from create_params, if this is a name, take
+    // it from the ID we get when creating the named disk.
+    let Some(boot_disk) =
+        params.create_params.boot_disk.as_ref().map(|x| x.name())
+    else {
+        return Ok(());
+    };
+
+    let instance_id = sagactx.lookup::<InstanceUuid>("instance_id")?;
+
+    let (.., authz_instance) = LookupPath::new(&opctx, &datastore)
+        .instance_id(instance_id.into_untyped_uuid())
+        .lookup_for(authz::Action::Modify)
+        .await
+        .map_err(ActionError::action_failed)?;
+
+    let (.., authz_disk) = LookupPath::new(&opctx, datastore)
+        .project_id(params.project_id)
+        .disk_name_owned(boot_disk.into())
+        .lookup_for(authz::Action::Read)
+        .await
+        .map_err(ActionError::action_failed)?;
+
+    let initial_configuration =
+        nexus_db_model::InstanceUpdate { boot_disk_id: Some(authz_disk.id()) };
+
+    datastore
+        .instance_reconfigure(&opctx, &authz_instance, initial_configuration)
+        .await
+        .map_err(ActionError::action_failed)?;
+
+    Ok(())
+}
+
+async fn sic_set_boot_disk_undo(
+    sagactx: NexusActionContext,
+) -> Result<(), anyhow::Error> {
+    let osagactx = sagactx.user_data();
+    let params = sagactx.saga_params::<Params>()?;
+    let datastore = osagactx.datastore();
+    let opctx = crate::context::op_context_for_saga_action(
+        &sagactx,
+        &params.serialized_authn,
+    );
+
+    let instance_id = sagactx.lookup::<InstanceUuid>("instance_id")?;
+
+    let (.., authz_instance) = LookupPath::new(&opctx, &datastore)
+        .instance_id(instance_id.into_untyped_uuid())
+        .lookup_for(authz::Action::Modify)
+        .await
+        .map_err(ActionError::action_failed)?;
+
+    // If there was a boot disk, clear it. If there was not a boot disk,
```

Comment:
Would it be too clever to try to look in the `params` to see if a boot disk was specified? Saves us from having to do the extra database calls to unset something we never bothered to set.